timemachine/cpp/src/bd_exchange_move.cu

```diff
@@ -0,0 +1,187 @@
+#include "bd_exchange_move.hpp"
+
+#include "constants.hpp"
+#include "gpu_utils.cuh"
+#include "kernels/k_exchange.cuh"
+#include "kernels/k_probability.cuh"
+#include "kernels/k_rotations.cuh"
+#include "math_utils.cuh"
+#include "mol_utils.hpp"
+
+namespace timemachine {
+
+template <typename RealType>
+BDExchangeMove<RealType>::BDExchangeMove(
+    const int N,
+    const std::vector<std::vector<int>> &target_mols,
+    const std::vector<double> &params,
+    const double temperature,
+    const double nb_beta,
+    const double cutoff,
+    const int seed,
+    const int proposals_per_move)
+    : N_(N), proposals_per_move_(proposals_per_move), num_target_mols_(target_mols.size()),
+      beta_(static_cast<RealType>(1.0 / (BOLTZ * temperature))), mol_potential_(N, target_mols, nb_beta, cutoff),
+      sampler_(num_target_mols_, seed), logsumexp_(N), d_intermediate_coords_(N * 3), d_params_(params.size()),
+      d_mol_energy_buffer_(num_target_mols_), d_mol_offsets_(get_mol_offsets(target_mols).size()),
+      d_log_weights_before_(num_target_mols_), d_log_weights_after_(num_target_mols_),
+      d_log_probabilities_before_(num_target_mols_), d_log_probabilities_after_(num_target_mols_),
+      d_log_sum_exp_before_(2), d_log_sum_exp_after_(2), d_samples_(1), d_quaternions_(round_up_even(4)),
+      d_translations_(round_up_even(4)), d_num_accepted_(1), num_attempted_(0) {
```

Comment:
`round_up_even_` here looks superfluous?

Reply:
Mostly for pedagogical purposes as I have been bitten by that a few times.

Reply:
sgtm